### PR TITLE
[NET-5772] Make tcp external service registered on terminating gw reachable from peered cluster

### PR DESCRIPTION
**Pull request by nathancoleman**
> [!NOTE]
> This is a followup to #18959 which fixed the same problem specifically when using `http` protocol.

### Description

The terminating gateways needs to be able to handle TCP connections from peers which are not TLS-terminated at the local mesh gateway. This amounts to including the target SNI that downstreams from peers would use when building the TLS context for the terminating gateway.

<details>
<summary>Filter chain match before this change</summary>

```yaml
filter_chain_match: {
  server_names: [
    "destination.default.dc2.internal.b81d6ee2-8454-768e-0a06-38a62620f76a.consul"
  ]
}
```
</details>

<details>
<summary>Filter chain match after this change</summary>

```yaml
filter_chain_match: {
  server_names: [
    "destination.default.dc2.internal.b81d6ee2-8454-768e-0a06-38a62620f76a.consul",
    "destination.default.default.dc1.external.b81d6ee2-8454-768e-0a06-38a62620f76a.consul"
  ]
}
```
</details>

### Testing & Reproduction steps
You can test this fix using the setup [here](https://github.com/nathancoleman/consul-lab/blob/main/k8s/terminating-gateway/multi-cluster/README.md), changing the `<ServiceDefaults>.spec.protocol` as necessary in `resources/dc2/external-service.yaml` if you want to see the behavior for `http` services as well.

You will need to pin a build of this branch as `global.image` in `values-dc2.yaml`. I would recommend doing an install without this build to witness things in their current broken state, and then doing a `helm upgrade` with `global.image` set.

### PR Checklist

* [x] updated test coverage
* [ ] external facing docs updated
* [x] appropriate backport labels added
* [x] not a security concern

### CODE REVIEW

**Comment by nathancoleman** [comment]
@erichaberkorn @jm96441n this still needs tests, a changelog entry, etc.; however, I'd appreciate it if you guys could sanity check my change here. I've verified that it works for `tcp` services and will verify that everything is what we expect with `http` service shortly.

File: agent/xds/listeners.go [hunk 1]
```diff
@@ -1188,29 +1190,9 @@ func createDownstreamTransportSocketForConnectTLS(cfgSnap *proxycfg.ConfigSnapsh
 	}
 
 	// Inject peering trust bundles if this service is exported to peered clusters.
-	if len(peerBundles) > 0 {
```
Comment (nathancoleman) [inline agent/xds/listeners.go:1191]: This logic gets moved out to a re-usable function below

**Comment by jm96441n** [comment]
approach overall LGTM, do you have an existing setup to run/test it out locally that I could steal to do some testing as well?

**Review by jm96441n — APPROVED** [review]
LGTM! nice work

**Pull request merged**